GH-15723: add `ReflectionConstant::getFileName()`

DanielEScherzer opened the pull request:
Allow determining the name of the file that defined a constant, when the constant was defined in userland code via `define()` or `const`. For constants defined by PHP core or extensions, `false` is returned, matching the existing `getFileName()` methods on other reflection classes.

dstogov reviewed (commented):
This adds new behavior in cost of increasing each zend_constant size.
New behavior requires the RFC process and can't go into the feature frozen PHP-8.4.
Personally, I don't see significant value. On the other hand, the cost is not significant.

janedbal commented:
> Personally, I don't see significant value.

One of the real-world usages I'd use it for is **unused composer package analysis**, see [current issue](https://github.com/shipmonk-rnd/composer-dependency-analyser/issues/137#issuecomment-2120538251)

iluuu1994 commented:
@DanielEScherzer Thank you for your PR! It's a bit late in the release process but you might still have time to consult the [internals mailing list](https://www.php.net/mailing-lists.php), and ask if there are any objections to adding this in 8.4.

> New behavior requires the RFC process and can't go into the feature frozen PHP-8.4.

This is no longer accurate. Small feature additions have historically been allowed without requiring RFCs. It's generally understood that they are ok if announced on the ML with no objections. This is indeed vague, and I think we would benefit from specifying this process. Additionally, features are now allowed all the way up to release candidates since https://wiki.php.net/rfc/release_cycle_update.

bwoebi commented:
I wonder whether it's reasonable to store the line number at the same time?

jorgsowa commented on ext/reflection/php_reflection.stub.php:
```diff
@@ -910,5 +910,7 @@ public function getValue(): mixed {}
 
     public function isDeprecated(): bool {}
 
+    public function getFileName(): string|false {}
```
There is no test case for the path where `getFileName` returns `false`. It would be useful to add such.

DanielEScherzer replied:
Its the first test case in `ReflectionConstant_getFileName.phpt` - the file name for `PHP_VERSION` is false because it is provided by PHP

jorgsowa replied:
That's true, but what do you think about improving the test coverage? You modified many places, thus paths, but provided only 4 test cases.

There are explicit changes to magic constants, com_dotnet ext, constants of different types, etc.

DanielEScherzer replied:
I think the implementation is pretty straightforward and that extra tests are not really needed

jorgsowa replied:
Full path coverage is always worth it. I don't understand your point.

DanielEScherzer replied:
I added tests for the other manually-registered constants, except for from the com_dotnet extension which I wasn't sure how to reach

DanielEScherzer commented:
I've sent an email to internals (my first one!) - https://news-web.php.net/php.internals/125571
If this gets approved I'll deal with the merge conflict

iluuu1994 commented:
There have been no objections on the ML, but unfortunately it has only been posted 5 days ago. 8.4 RC1 is tagged on Tuesday, so I would suggest delaying this for 8.5.

DanielEScherzer commented:
> There have been no objections on the ML, but unfortunately it has only been posted 5 days ago. 8.4 RC1 is tagged on Tuesday, so I would suggest delaying this for 8.5.

Thats understandable, but annoying that this won't be included. Can we add the `Waiting on RM` label and see if they might be willing to include it? @php/release-managers-84

iluuu1994 commented:
@janedbal No, RC1 has landed. This will be delayed to 8.5.

DanielEScherzer commented:
Well, this missed 8.4, but since that has been branched I hope it can be merged for 8.5 - I've rebased to resolve the merge conflict and added an entry to UPGRADING

iluuu1994 reviewed (commented):
It seems a bit inconsistent not to offer `getStartLine` (`getEndLine` might be overkill).

Another note: We don't have this field or `getStartLine` on `ReflectionClassConstant`. The former we can get through `getDeclaringClass` though.

DanielEScherzer commented on ext/reflection/tests/ReflectionConstant_getFileName.phpt:
```diff
@@ -0,0 +1,44 @@
+const IN_CURRENT_FILE_AST = 123;
+
+echo "From PHP:\n";
+testConstant( 'PHP_VERSION' );
```
force of habit; done

DanielEScherzer commented:
> It seems a bit inconsistent not to offer `getStartLine` (`getEndLine` might be overkill).
> 
> Another note: We don't have this field or `getStartLine` on `ReflectionClassConstant`. The former we can get through `getDeclaringClass` though.

I'd be happy to add `getStartLine`, but it means that we need to track the line where every constant is declared - I figured that without someone asking for it / having a use case, it would just add extra memory to every constant PHP tracks

iluuu1994 commented on Zend/zend_builtin_functions.c:
```diff
@@ -549,6 +549,13 @@ ZEND_FUNCTION(define)
+		c.filename = NULL;
+	} else {
+		c.filename = zend_string_copy(filename);
+	}
```
Any updates on this?

DanielEScherzer replied:
Done
